core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -586,37 +591,51 @@ protected StructReader(List<ValueReader<?>> readers, Schema schema) {
       }
 
       if (isDeletedColumnPos == null) {
-        this.positions = new int[0];
-        this.constants = new Object[0];
+        this.constantPositions = new int[0];
+        this.constantValues = new Object[0];
       } else {
-        this.positions = new int[] {isDeletedColumnPos};
-        this.constants = new Object[] {false};
+        this.constantPositions = new int[] {isDeletedColumnPos};
+        this.constantValues = new Object[] {false};
       }
+      this.defaultPositions = new int[0];
+      this.defaultValues = new Object[0];
     }
 
     protected StructReader(
         List<ValueReader<?>> readers, Types.StructType struct, Map<Integer, ?> idToConstant) {
       this.readers = readers.toArray(new ValueReader[0]);
 
       List<Types.NestedField> fields = struct.fields();
-      List<Integer> positionList = Lists.newArrayListWithCapacity(fields.size());
-      List<Object> constantList = Lists.newArrayListWithCapacity(fields.size());
+      List<Integer> constantPositionsList = Lists.newArrayListWithCapacity(fields.size());
+      List<Object> constantValuesList = Lists.newArrayListWithCapacity(fields.size());
+      List<Integer> defaultValuesPositionList = Lists.newArrayListWithCapacity(fields.size());
+      List<Object> defaultValuesList = Lists.newArrayListWithCapacity(fields.size());
       for (int pos = 0; pos < fields.size(); pos += 1) {
         Types.NestedField field = fields.get(pos);
         if (idToConstant.containsKey(field.fieldId())) {
-          positionList.add(pos);
-          constantList.add(idToConstant.get(field.fieldId()));
+          constantPositionsList.add(pos);
+          constantValuesList.add(idToConstant.get(field.fieldId()));
         } else if (field.fieldId() == MetadataColumns.ROW_POSITION.fieldId()) {
           // track where the _pos field is located for setRowPositionSupplier
           this.posField = pos;
         } else if (field.fieldId() == MetadataColumns.IS_DELETED.fieldId()) {
-          positionList.add(pos);
-          constantList.add(false);
+          constantPositionsList.add(pos);
+          constantValuesList.add(false);
+        } else if (field.initialDefault() != null) {
+          // Add a constant value for fields that have a default value.
+          // In the {@link #read()} method, this will be leveraged only if there is no corresponding
+          // reader.
+          defaultValuesPositionList.add(pos);
+          defaultValuesList.add(
+              IdentityPartitionConverters.convertConstant(field.type(), field.initialDefault()));
```

Comment:
This fixes the problem for Iceberg generics, but this conversion is not valid for other in-memory data models, like Spark or Flink. Those also extend this class.

We either need to pass in the conversion method or pass in a map of default constants like the `idToConstant` map above. I think I'd prefer to pass in an `idToDefault` map that is used in the same way and produced from the schema.

Using a map will also help solve the problem of structs with nested defaults.

Reply:
Also, in order to make sure that we are not producing incorrect results for the other object models, we will need to fail if a default value is needed by this class, but is unavailable.